File: pingpong/src/pingpong.c
/* Compile under Linux with "cc pingpong.c -o pingpong -lpthread; ./pingpong"
Does not run under macOS as unnamed semaphores with sem_init are not supported.

Unnamed pthread semaphores are used for passing the "ball" between a ping thread
and a pong thread. Two split binary semaphores (i.e. at most one of them is 1)
are used for synchronization and for mutually exclusive access to a global
variable with the number of remaining bounces of the ball. The ping and pong
threads "do not know" about each other, they only communicate via the semaphores
and the global variable.

The main program waits for the ping thread to terminate, which it does when the
number of remaining bounces reaches 0.
*/
#include <pthread.h>
#include <stdio.h>
#include <semaphore.h>
#include <stdlib.h>

// #define Rounds 100000

sem_t ping_mutex, pong_mutex;
int bounces = 0;

void *ping(void *argptr) {
  sem_wait(&ping_mutex);
  while (bounces > 0) {
    bounces -= 1;
    sem_post(&pong_mutex);
    sem_wait(&ping_mutex);
  }
  // printf("bounces: %d\n", bounces);
}
void *pong(void *argptr) {
  sem_wait(&pong_mutex);
  while (bounces > 0) {
    bounces -= 1;
    sem_post(&ping_mutex);
    sem_wait(&pong_mutex);
  }
}
int main(int argc, char *argv[]) {
  if(argc < 2){
      printf("The Usage: %s Rounds\n", argv[0]);
      exit(1);
  }
  
  bounces = atoi(argv[1]);
  pthread_t ping_tid, pong_tid;
  sem_init(&ping_mutex, 0, 1);
  sem_init(&pong_mutex, 0, 0);
  pthread_create(&ping_tid, NULL, ping, NULL);
  pthread_create(&pong_tid, NULL, pong, NULL);
  pthread_join(ping_tid, NULL);
  return 0;
}

File: pingpong/src/pingpong.java
/* Compile and run under Linux and macOS with "javac pingpong.java; java Start"

Java monitors with signalling by notify and wait are used. Classes Ping and Pong
define active objects in the sense that they have methods that can be called and
each has a thread that calls methods of other objects. The thread and the
methods of each object are synchronized, i.e. run mutually exclusive. Two
objects are created: the Ping object calls the Pong object and passes itself as
a parameter, the Pong object asynchronously calls back the Ping object. The
number of bounces of the "ball" is passed along as well.

In Java, a program terminates when all threads terminate. The Ping and Pong
threads terminte when the number of bounces reaches 0 and 1, respectively.
*/
class Ping extends Thread {
  private Pong other;
  private int bounces = 0;
  Ping(Pong o, int b) {
    other = o; bounces = b;
  }
  synchronized void ping(int b) {
    bounces = b; notify();
  }
  synchronized public void run() {
    while (bounces > 0) {
      other.pong(this, bounces - 1);
      try {wait();} catch (InterruptedException e) {}
    }
  }
}
class Pong extends Thread {
  private Ping other;
  private int bounces;
  synchronized void pong(Ping o, int b) {
    other = o; bounces = b; notify();
  }
  synchronized public void run() {
    do {
      try {wait();} catch (InterruptedException e) {}
      other.ping(bounces - 1);
    } while (bounces > 1);
  }
}
class PingPong  {
  //static final int Rounds = 100000;
  public static void main(String args[]) {
    if(args.length < 1){
        System.err.println("Usage: java PingPong Rounds\n");
        return;
    }
    int Rounds = Integer.parseInt(args[0]);
    Pong pong = new Pong();
    Ping ping = new Ping(pong, Rounds);
    pong.start(); ping.start();
  }
}
